src/NuGet.Core/NuGet.Build.Tasks/StaticGraphRestoreTaskBase.cs

```diff
@@ -97,34 +103,75 @@ public override bool Execute()
                 using (var loggingQueue = new TaskLoggingQueue(Log))
                 using (var process = new Process())
                 {
+                    bool errorLogged = false;
+
                     process.EnableRaisingEvents = true;
                     process.StartInfo = new ProcessStartInfo
                     {
-                        Arguments = $"\"{string.Join("\" \"", GetCommandLineArguments())}\"",
+                        Arguments = GetCommandLineArguments(globalProperties),
                         CreateNoWindow = true,
                         FileName = GetProcessFileName(ProcessFileName),
+                        RedirectStandardError = true,
                         RedirectStandardInput = true,
                         RedirectStandardOutput = true,
+#if !NETFRAMEWORK
+                        StandardInputEncoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false),
+#endif
                         UseShellExecute = false,
                         WorkingDirectory = Environment.CurrentDirectory,
                     };
 
                     // Place the output in the queue which handles logging messages coming through on StdOut
                     process.OutputDataReceived += (sender, args) => loggingQueue.Enqueue(args?.Data);
 
+                    process.ErrorDataReceived += (_, args) =>
+                    {
+                        if (args.Data != null)
+                        {
+                            // Any message on the standard error stream should be logged as an error
+                            Log.LogError(args.Data);
+
+                            errorLogged = true;
+                        }
+                    };
+
                     process.Exited += (sender, args) => semaphore.Release();
 
                     try
                     {
                         Log.LogMessageFromResources(MessageImportance.Low, nameof(Strings.Log_RunningStaticGraphRestoreCommand), process.StartInfo.FileName, process.StartInfo.Arguments);
 
-                        process.Start();
+                        Encoding previousConsoleInputEncoding = Console.InputEncoding;
 
-                        process.BeginOutputReadLine();
+                        // Set the input encoding to UTF8 without a byte order mark, the spawned process will use this encoding on .NET Framework
+                        Console.InputEncoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false);
+
+                        try
+                        {
+                            process.Start();
+
+                            process.BeginOutputReadLine();
+                            process.BeginErrorReadLine();
+
+                            if (SerializeGlobalProperties)
+                            {
+                                using var writer = new BinaryWriter(process.StandardInput.BaseStream, Encoding.UTF8, leaveOpen: true);
```

Comment:
Why `Encoding.UTF8` and not `new UTF8Encoding(encoderShouldEmitUTF8Identifier: false)`?

[`Encoding.UTF8` passes `true` for `encoderShouldEmitUTF8Identifier`](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Text/UTF8Encoding.cs,62), which implies that [we're going to write a BOM when we don't expect it](https://github.com/NuGet/NuGet.Client/pull/5413/files#diff-bbc3e4c20edf3a20b032ba04bfe636becd616a9455d593f9078e8bc33a33244fR152-R153).

If my understanding is correct, why wasn't this caught by a test?

Reply:
As far as I know based on reading the docs and looking at the source code, the Encoding specified to the `BinaryWriter`/`BinaryReader` is only used to determine the size of a character (1 byte ASCII vs 2 byte Unicode).  Since the stream is just a collection of bytes, these classes don't do anything with an encoding preamble/byte order mark.  Since we own the code that writes and reads the data, we just need to tell the reader and the writer that we're using UTF8 for characters.  

The `TextWriter`/`TextReader` do emit and read the preamble/byte order mark since those streams will represent a set of text characters and the reader will need to know how that from reading the preamble.